File: application/app/domain/model/Authentication.scala
package domain.model

import common.validation.Validatable
import javax.validation.constraints.{NotBlank, Size, NotNull}
import org.hibernate.validator.constraints.URL

import scala.annotation.meta.field

sealed trait Authentication extends Validatable {

}

object Authentication {
  case class None() extends Authentication

  case class Basic(
                    @(NotBlank @field) @(Size @field)(min = 5, max = 255) username: String,
                    @(NotBlank @field) @(Size @field)(min = 5, max = 255) password: String
                  ) extends Authentication

  case class Bearer(
                     @(NotNull @field) @(URL @field) @(Size @field)(min = 5, max = 255) verifyTokenUrl: String
                   ) extends Authentication
}

File: ws-gateway-domain/src/main/java/com/cosmin/wsgateway/domain/Authentication.java
package com.cosmin.wsgateway.domain;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import lombok.Value;
import org.hibernate.validator.constraints.URL;


public interface Authentication {
    class None implements Authentication {

    }

    @Value
    class Basic implements Authentication {
        @NotBlank
        @Size(min = 5, max = 255)
        private final String username;

        @NotBlank
        @Size(min = 5, max = 255)
        private final String password;
    }

    @Value
    class Bearer implements Authentication {
        @NotNull
        @URL
        private final String authorizationServerUrl;
    }
}
